### PR TITLE
Introduce retries to some git operations

### PR DESCRIPTION
**Pull request by puerco**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR introduces the capability to retry git operations when certain errors occur.  It inctroduces a new wrapper `NetworkError()` which extends the standard golang error with a method `CanRetry()` that checks the error message to infer if we're dealing with a network error that can be retried.

To enable retries, a program can call the `repo.SetMaxRetries()` function passing the number of retries to attempt when dealing with a network error.

In preparation for #1595, the following functions now hab retries enabled:

- `repo.HasRemoteBranch()`
- `repo.Push()`

It also adds some initial tests for the new functions and error type.

#### Which issue(s) this PR fixes:

Related to #1595

#### Special notes for your reviewer:

Depends on #1620 to run

#### Does this PR introduce a user-facing change?

```release-note
- `git.Repo()` now has a `SetMaxRetries()` function to enable retrying git ops when a network error occurs.
- New function `repo.HasBranch()` added to the git package
- Added a `NetworkError` type that wraps the golang standard error and can be queried to check if errors can be retried
- `git.Push()` and `git.HasRemoteBranch()` now have retry capability
```

### CODE REVIEW

File: pkg/git/git.go [hunk 1]
```diff
@@ -587,7 +610,25 @@ func (r *Repo) Push(remoteBranch string) error {
...
 	args = append(args, DefaultRemote, remoteBranch)
 
-	return command.NewWithWorkDir(r.Dir(), gitExecutable, args...).RunSuccess()
+	if err := command.NewWithWorkDir(r.Dir(), gitExecutable, args...).RunSuccess(); err != nil {
+		return NewNetworkError(err)
+	}
+	return nil
+}
+
+// PushWithRetries tries to push a branch, retrying when the error allows it
```
Comment (hasheddan) [inline pkg/git/git.go:619]: Really cool to see this new functionality @puerco! One thing I was wondering was how often we think there may be additional `___WithRetries` methods? If we think there will be more than this one (seems likely), what would you think about making the retry configuration part of `Repo` itself, similar to what we do with `dryRun`, etc.?

Comment (puerco) [reply pkg/git/git.go:619]: Good point @hasheddan , I'm going to modify it that way and we can start adding retries to the git functions as we need them :+1:

File: pkg/git/git.go [hunk 2]
```diff
@@ -588,7 +610,25 @@ func (r *Repo) Push(remoteBranch string) error {
...
+		logrus.Debugf("Pushing %s, (will retry %d times)", remoteBranch, numRetries)
+		if err = r.Push(remoteBranch); err == nil {
+			return nil
+		}
+
+		if !err.(NetworkError).CanRetry() {
+			return errors.Wrap(err, "cannot retry")
+		}
+	}
+	return errors.New(fmt.Sprintf("still could not push %s after %d retries", remoteBranch, numRetries))
```
Comment (saschagrunert) [inline pkg/git/git.go:631]:
```suggestion
	return errors.Errorf("still could not push %s after %d retries", remoteBranch, numRetries)
```

**Review by saschagrunert — COMMENTED** [review]
Just one nit from my side. 👇

File: pkg/git/git.go (hunk 2 above)
Comment (puerco) [reply pkg/git/git.go]: Done (line is gone now)

File: pkg/git/git.go (hunk 1 above)
Comment (puerco) [reply pkg/git/git.go]: OK, I've added the retry capability to the `git.Repo` object. It now has a new method to define the number of retries:

`repo.SetMaxRetries()`

**Comment by puerco** [comment]
Retries are now built into the `git.Push()` and `git.HasRemoteBranch()` methods. A couple of network errors are now handled (would be useful to collect more retry-worthy cases).

When hitting a network error, it will retry up to `repo.maxRetries` times, exponentially backing off after each:

```
$ sudo route add -net 140.82.0.0/16 reject
$ go run cmd/krel/main.go anago push-git-objects --repo=/tmp/testPush/ --release-branch=release-1.19 --max-retries=5 
INFO Checkout master branch to push objects       
Pausing here. Confirm push --dry-run of tags and bits (1/10) 
y
INFO Pushing --dry-run 0 tags                     
INFO Pushing --dry-run release-1.19 branch:       
INFO Verifying release-1.19 branch exists in the repo 
INFO Branch release-1.19 found in the repository  
INFO Verifying release-1.19 branch exists on the remote 
WARN Could not list references on the remote repository. 
ERRO Error listing remote references (will retry 5 times in 0 secs): dial tcp 140.82.113.3:22: connect: no route to host 
WARN Could not list references on the remote repository. 
ERRO Error listing remote references (will retry 4 times in 1 secs): dial tcp 140.82.113.3:22: connect: no route to host 
WARN Could not list references on the remote repository. 
ERRO Error listing remote references (will retry 3 times in 2 secs): dial tcp 140.82.114.3:22: connect: no route to host 
WARN Could not list references on the remote repository. 
ERRO Error listing remote references (will retry 2 times in 4 secs): dial tcp 140.82.114.3:22: connect: no route to host 
WARN Could not list references on the remote repository. 
ERRO Error listing remote references (will retry 1 times in 8 secs): dial tcp 140.82.113.3:22: connect: no route to host 
WARN Could not list references on the remote repository. 
FATA pushing branch release-1.19: checking if branch release-1.19 exists in remote repository: dial tcp 140.82.114.3:22: connect: no route to host 
exit status 1

```

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Pull request merged**